src/providers/Web3Provider.tsx

```diff
@@ -60,36 +60,101 @@ export const Web3ContextProvider: FC<PropsWithChildren> = ({ children }) => {
     ...web3ProviderInitialState,
   })
 
+  const _connectionChanged = (isConnected: boolean) => {
+    setState(prevState => ({
+      ...prevState,
+      isConnected,
+    }))
+  }
+
+  const _accountsChanged = useCallback((accounts: string[]) => {
+    if (accounts.length <= 0) {
+      _connectionChanged(false)
+      return
+    }
+
+    const [account] = accounts
+    setState(prevState => ({
+      ...prevState,
+      account,
+    }))
+  }, [])
+
+  const _setNetworkSpecificVars = (
+    chainId: number,
+    sapphireEthProvider = state.sapphireEthProvider!,
+  ): void => {
+    if (!sapphireEthProvider) {
+      throw new Error('[Web3Context] Sapphire provider is required!')
+    }
+
+    if (!(chainId in NETWORKS)) {
+      throw new UnknownNetworkError('Unknown network!')
+    }
+
+    const { wRoseContractAddress, explorerBaseUrl, networkName } = NETWORKS[chainId]
+
+    const wRoseContract = new ethers.Contract(
+      wRoseContractAddress,
+      WrappedRoseMetadata.output.abi,
+      sapphireEthProvider.getSigner(),
+    )
+
+    setState(prevState => ({
+      ...prevState,
+      wRoseContract,
+      explorerBaseUrl,
+      networkName,
+      wRoseContractAddress,
+    }))
+  }
+
+  const _chainChanged = useCallback(() => {
+    window.location.reload()
+  }, [])
+
+  // TODO: This should probably use separate status, like isOffline(with warning message), to not interrupt the user flow
+  const _connect = useCallback(() => _connectionChanged(true), [])
+  const _disconnect = useCallback(() => _connectionChanged(false), [])
+
+  const _addEventListeners = (ethProvider = window.ethereum!) => {
+    ethProvider.on('accountsChanged', _accountsChanged)
+    ethProvider.on('chainChanged', _chainChanged)
+    ethProvider.on('connect', _connect)
+    ethProvider.on('disconnect', _disconnect)
+  }
+
+  const _removeEventListeners = (ethProvider = window.ethereum!) => {
+    ethProvider.off('accountsChanged', _accountsChanged)
+    ethProvider.off('chainChanged', _chainChanged)
+    ethProvider.off('connect', _connect)
+    ethProvider.off('disconnect', _disconnect)
+  }
+
   const _init = async (account: string) => {
+    _removeEventListeners()
+
     try {
-      const ethProvider = new ethers.providers.Web3Provider(window.ethereum)
+      const ethProvider = new ethers.providers.Web3Provider(window.ethereum!)
       const sapphireEthProvider = sapphire.wrap(ethProvider) as ethers.providers.Web3Provider &
         sapphire.SapphireAnnex
 
       const network = await sapphireEthProvider.getNetwork()
 
-      if (!(network.chainId in NETWORKS)) {
-        return Promise.reject(new UnknownNetworkError('Unknown network!'))
+      try {
+        await _setNetworkSpecificVars(network.chainId, sapphireEthProvider)
+      } catch (ex) {
+        return Promise.reject(ex)
```

Comment:
This is odd. Is it not the same as `throw ex`. And the same as removing `try catch` all together

Reply:
So idea behind this was, that `_setNetworkSpecificVars` throws specific `UnknownNetworkError` which I want to propagate to the caller. The whole `_init` is inside another try-catch block that rethrows `Unable to initialize providers` generic message. So either I rethrow the thrown error in `_init`, or return the Promise with the error(The way it is now).

Reply:
Ah. Ew. Just add `if (ex instanceof UnknownNetworkError) throw ex` to the upper catch